### PR TITLE
Add SeRoNet Spawner

### PR DESCRIPTION
**Pull request by mael25**
This project enables to use a REST API to spawn different kind of boxes. It comes with two worlds. The first is a demo world (`spawner_demo.wbt`) where simple demo controllers are already implemented. The second , `spawner_seronet.wbt`, is meant to be used with the [SeRoNet-Tooling-Collection](https://xito.one), so all the controller are `<extern>` (except the inner controller of the `ConveyorStation`, `box_creator`, which handles the REST API and is independent).

**Dependencies**: https://github.com/microsoft/cpprestsdk

### CODE REVIEW

File: projects/seronet-spawner/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# SeRoNet Spawner
+![spawner_small](https://user-images.githubusercontent.com/38250944/123643493-b8748b80-d824-11eb-9a5a-6d17e9b63180.png)
+
+
+This project enables to use a REST API to spawn different kind of boxes, such as `CardboardBox` or `PlasticCrate`. It comes with two worlds: the first is a demo world (`spawner_demo.wbt`) where the different controllers are already implemented. The second, `spawner_seronet.wbt`, is meant to be used with the SeRoNet-Tooling-Collection software, so all the controllers (robot, conveyor belts, etc.) are defined as `<extern>` (except the inner controller of the `ConveyorStation`, `box_creator`, which handles the REST API and is independent).
```
Comment (omichel) [inline projects/seronet-spawner/README.md]: One sentence per line:
```suggestion
This project enables to use a REST API to spawn different kind of boxes, such as `CardboardBox` or `PlasticCrate`.
It comes with two worlds: the first is a demo world (`spawner_demo.wbt`) where the different controllers are already implemented.
The second, `spawner_seronet.wbt`, is meant to be used with the SeRoNet-Tooling-Collection software, so all the controllers (robot, conveyor belts, etc.) are defined as `<extern>` (except the inner controller of the `ConveyorStation`, `box_creator`, which handles the REST API and is independent).
```

File: projects/seronet-spawner/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+## Configuration
+
+Both worlds make use of a `ConveyorStation` with 2 `ConveyorStationPlatform`: one is the place where robots deposit boxes to remove, while the other outputs the box requested by the REST API.
+
+In general though, the `ConveyorStation` can have an adjustable number of `ConveyorStationPlatform` (by setting `platformNumber`). The name of the components for the `i`th `ConveyorStationPlatform` is:
```
Comment (omichel) [inline projects/seronet-spawner/README.md]:
```suggestion
In general though, the `ConveyorStation` can have an adjustable number of `ConveyorStationPlatform` (by setting `platformNumber`).
The name of the components for the `i`th `ConveyorStationPlatform` is:
```

File: projects/seronet-spawner/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+* Left motor: `motorNamePrefix`_`i`_left
+* Right motor: `motorNamePrefix`_`i`_right
+* Distance sensor: `sensorNamePrefix`_`i`
+
+If there is only one `ConveyorStationPlatform` the names are simply (`motorNamePrefix`_left, `motorNamePrefix`_right, `sensorNamePrefix`). In our case, the platform n°1 is the removal place and n°2 the spawning place.
```
Comment (omichel) [inline projects/seronet-spawner/README.md]: ° is French:
```suggestion
If there is only one `ConveyorStationPlatform` the names are simply (`motorNamePrefix`_left, `motorNamePrefix`_right, `sensorNamePrefix`). In our case, the platform number 1 is the removal place and number 2 the spawning place.
```

File: projects/seronet-spawner/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+* `objectId`: the indivudal identifiant (integer) to give to the box.
+* values:
+  * `typeId`: the type of box to spawn (see **Edit box types** section below)
+  * `size`: the size of the box in meters (LxHxW)
+
+### Edit box types
+`typeId`, the type of box to spawn, search for the corresponding predefined types in `controllers/box_creator/utils/boxtypes.json` that can be edited or completed.
+
+## Additional information for SeRonet-Tooling-Collection
+The main controller of the ConveyorStation is for the OPC_UA interface. An updated version (draft, not tested) of the OPC_UA_STATION controller with the new names of components, `Spawner_OPC_UA_STATION`, is set as default for the `spawner_seronet.wbt` world. The existing OPC_UA motor commands are applied to the conveyor belt of the remover part of the `ConveyorStation`. The motor of the spawner part are initialized in speed control but the interface with OPC_UA must be implemented.
```
Comment (omichel) [inline projects/seronet-spawner/README.md]:
```suggestion
The main controller of the ConveyorStation is for the OPC_UA interface.
An updated version (draft, not tested) of the OPC_UA_STATION controller with the new names of components, `Spawner_OPC_UA_STATION`, is set as default for the `spawner_seronet.wbt` world.
The existing OPC_UA motor commands are applied to the conveyor belt of the remover part of the `ConveyorStation`.
The motor of the spawner part are initialized in speed control but the interface with OPC_UA must be implemented.
```

File: projects/seronet-spawner/controllers/box_creator/box_creator.cc [hunk 5]
```diff
@@ -0,0 +1,144 @@
+/*
```
Comment (omichel) [inline projects/seronet-spawner/controllers/box_creator/box_creator.cc]: We normally use the `.cpp` extension for C++ file, not `.cc`

File: projects/seronet-spawner/controllers/box_creator/utils/BoxType.cc [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#include "BoxType.hh"
+#include <algorithm>
+
+BoxType::BoxType(string proto, double *rotation, double *size) : proto(proto) {
+  for (int i = 0; i < 4; i++) {
+    this->rotation[i] = rotation[i];
+  }
+  for (int i = 0; i < 3; i++) {
+    this->size[i] = size[i];
+  }
```
Comment (omichel) [inline projects/seronet-spawner/controllers/box_creator/utils/BoxType.cc]: Coding style is wrong:
```suggestion
  for (int i = 0; i < 4; i++)
    this->rotation[i] = rotation[i];
  for (int i = 0; i < 3; i++)
    this->size[i] = size[i];
```

File: projects/seronet-spawner/controllers/box_creator/utils/CoordinationRESTServer.cc [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+  // extract the json value from the request object
+  auto jvalue = request.extract_json().get();
+
+  if (resource == "addObject") {
+    response.set_status_code(handle_add_new_object(jvalue));
+  } else {
+    // error code 404: resource not found
+    response.set_status_code(web::http::status_codes::NotFound);
+  }
```
Comment (omichel) [inline projects/seronet-spawner/controllers/box_creator/utils/CoordinationRESTServer.cc]:
```suggestion
  if (resource == "addObject")
    response.set_status_code(handle_add_new_object(jvalue));
  else  // error code 404: resource not found
    response.set_status_code(web::http::status_codes::NotFound);
```

File: projects/seronet-spawner/controllers/box_creator/utils/CoordinationRESTServer.cc [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  } catch (json::json_exception &ex) {
+    std::cerr << "Addobject: Error in request syntax (" << ex.what() << ")"
+              << endl;
+    return 0;
+  }
+
+  bool addOK = true;
+  if (addOK) {
+    return status_codes::OK;
+  }
```
Comment (omichel) [inline projects/seronet-spawner/controllers/box_creator/utils/CoordinationRESTServer.cc]:
```suggestion
  if (addOK)
    return status_codes::OK;
```

File: projects/seronet-spawner/protos/Kasten.proto [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+PROTO Kasten [
+  field SFVec3f    translation         0 0 0
+  field SFRotation rotation            0 1 0 0
+  field SFString   name                "Kasten"
+  field SFVec3f    size                0.4 0.12 0.3                  # Defines the size of the shelf
+  field SFColor    baseColor           0.447059 0.623529 0.811765
+  
+  field SFFloat    plateWidth          0.016                         # Defines the frame of the plate
+
```
Comment (omichel) [inline projects/seronet-spawner/protos/Kasten.proto]:
```suggestion
  field SFFloat    plateWidth          0.016                         # Defines the frame of the plate
```

File: projects/seronet-spawner/protos/Kasten.proto [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+  
+  field SFString   contactMaterial     "default"                     # any string
+  field SFFloat    mass                0.92                         # Defines the mass of the box in kg.
+  field SFBool     locked              FALSE
+]
+{
+  %{
+    local wbrotation = require('wbrotation')
+
+    -- check inputs
```
Comment (omichel) [inline projects/seronet-spawner/protos/Kasten.proto]:
```suggestion
    -- check input
```

**Review by omichel — CHANGES_REQUESTED** [review]
We usually use the `.cpp` and `.hpp` file extensions for C++.
However if these files were simply copied / pasted from somewhere else, maybe we can keep them as is.

File: projects/seronet-spawner/controllers/box_creator/box_creator.cc (hunk 5 above)
Comment (mael25) [reply projects/seronet-spawner/controllers/box_creator/box_creator.cc]: > We usually use the `.cpp` and `.hpp` file extensions for C++.
> However if these files were simply copied / pasted from somewhere else, maybe we can keep them as is.

The files in SeRoNet use `.cc` and `.hh`, but only `CoordinationRESTServer` was copied from there, so I think we can change the extensions...

**Review by omichel — APPROVED** [review]
Thank you.

**Pull request merged**